platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -486,4 +497,61 @@ public ValueProducer getProducer() {
             return left.getProducer().plus(right.getProducer());
         }
     }
+
+    protected CollectionSupplierGuard<T, C> guardSupplier(CollectionSupplier<T, C> supplier) {
+        return new CollectionSupplierGuard<>(this, supplier);
+    }
+
+    protected static final class CollectionSupplierGuard<T, C extends Collection<T>> implements CollectionSupplier<T, C>, GuardedData<CollectionSupplier<T, C>> {
```

Comment:
`GuardedSupplier` would read better IMO (for this and `MapProperty`'s, which could use the same name).

Reply:
>GuardedSupplier would read better IMO (for this and MapProperty's, which could use the same name).

It is indeed, a supplier, but we're losing symmetry with the `ProviderGuard` (which isn't a provider, so this trick won't work there).